File: src/main/java/facts/stmt/CJmpStmt.java
package facts.stmt;

import java.util.Collections;
import java.util.List;
import facts.exp.Expr;
import facts.exp.Var;

/**
 * Conditional Jump fact
 */
public class CJmpStmt extends Stmt {

    private String target;
    private Expr condition; // TODO changed this to ExpFact but should check this

    public CJmpStmt(String pc, String target, Var condition) {
        super(pc);
        this.target = target;
        this.condition = condition;
    }

    public String getTarget() {
        return target;
    }

    public Expr getCondition() {
        return condition;
    }

    @Override
    public String toString() {
        return String.format("%sif (%s) goto label%s;", getLabel(), condition, target);
    }

    @Override
    public List<Expr> getChildren() {
        return Collections.singletonList(condition);
    }

    @Override
    public void replace(Expr oldExp, Expr newExp) {
        if (condition.equals(oldExp)) {
            condition = newExp;
        }
    }
}

File: src/main/scala/astnodes/stmt/CJmpStmt.scala
package astnodes.stmt

import astnodes.exp.Expr
import astnodes.exp.variable.{Register, Variable}

/** Conditional Jump fact
  */
case class CJmpStmt(override val pc: String,
  trueTarget: String,
  falseTarget: String,
  var condition: Expr,
) extends Stmt(pc) {
  def getCondition: Expr = condition
  override def toString = s"if ($condition) goto label$trueTarget else goto label$falseTarget;"
  override def toBoogieString: String = s"if (bv1tobool(${condition.toBoogieString})) { goto label$trueTarget; } goto label$falseTarget;"

  override def subst(v: Variable, w: Variable): Stmt = this.copy(condition = condition.subst(v,w))
  def fold(old: Expr, sub: Expr): Stmt = this.copy(condition = condition.fold(old, sub))
}
